tests/BaseElementTest.php

```diff
@@ -236,4 +252,141 @@ public function testOnBeforeWriteNoParent()
 
         $this->assertEquals(0, (int) $element1->Sort);
     }
+
+    public function getElementCMSLinkDataProvider()
+    {
+        return [
+            // Element in DataObject with $directLink === true
+            'element1' => [
+                TestElement::class,
+                'elementDataObject1',
+                'http://localhost/test/1/ItemEditForm/field/ElementalArea/item/',
+                true
+            ],
+            // Element in DataObject with $inline_editable = false
+            'element2' => [
+                TestElementDataObject::class,
+                'testElementDataObject1',
+                'http://localhost/test/1/ItemEditForm/field/ElementalArea/item/',
+            ],
+            // Element in DataObject with $inline_editable = true
+            'element3' => [
+                ElementContent::class,
+                'contentDataObject1',
+                'http://localhost/test/1/ItemEditForm',
+            ],
+            // Element in Page with $inline_editable = true
+            'element4' => [
+                ElementContent::class,
+                'content1',
+                'http://localhost/admin/pages/edit/show/1',
+            ],
+            // Element in DataObject with $directLink === true
+            'element5' => [
+                ElementContent::class,
+                'content1',
+                'admin/pages/edit/EditForm/1/field/ElementalArea/item/1/edit',
+                true
+            ],
+            // Element without any Page or DataObject
+            'element6' => [
+                ElementContent::class,
+                'contentDataObject2',
+                null
+            ],
+            // DataObject without CMSEditLink method
+            'element7' => [
+                TestElement::class,
+                'elementDataObject2',
+                null
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider getElementCMSLinkDataProvider
+     */
+    public function testCMSEditLink(string $class, string $element, ?string $link, bool $directLink = false)
+    {
+        $object = $this->objFromFixture($class, $element);
+        $editLink = $object->CMSEditLink($directLink);
+
+        if ($link) {
+            $this->assertStringContainsString($link, $editLink);
+        } else {
+            $this->assertNull($link);
+        }
+    }
+
+    public function canDeleteProvider()
+    {
+        return [
+            // Element on Page
+            'element1' => [
+                ElementContent::class,
+                'content1',
+                'ADMIN',
+            ],
+            // Element in DataObject
+            'element2' => [
+                TestElement::class,
+                'elementDataObject1',
+                'CMS_ACCESS_CMSMain',
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider canDeleteProvider
+     */
+    public function testCanDelete(
```

Comment:
This checks that the result is not null, but it doesn't validate that the method works as expected - for example this won't confirm that a user with `canDelete()` permissions on the underlying `DataObject` has `canDelete()` permission on the element, nor (arguably more importantly) that users _without_ permission on the object don't have permission on the element.

Reply:
Actually, the main purpose of this method check that it invoke canDelete or canArchive methods in class and don't return NULL or throws Exception. I don't have an aim to check permissions.  
DONE!